mm2src/mm2_main/src/rpc/dispatcher/dispatcher.rs

```diff
@@ -163,6 +163,7 @@ async fn dispatcher_v2(request: MmRpcRequest, ctx: MmArc) -> DispatcherResult<Re
         "enable_slp" => handle_mmrpc(ctx, request, enable_token::<SlpToken>).await,
         "enable_eth_with_tokens" => handle_mmrpc(ctx, request, enable_platform_coin_with_tokens::<EthCoin>).await,
         "enable_erc20" => handle_mmrpc(ctx, request, enable_token::<EthCoin>).await,
+        "enable_nft" => handle_mmrpc(ctx, request, enable_token::<EthCoin>).await,
```

Comment:
Aren't those basically the same method but with 2 different aliases, e.g. if we passed erc20 activation params to `enable_nft` it will work. Doesn't this open the possibility for API users to make undetectable mistakes?
We probably need to make the `enable_token` function more generic, e.g.
```rust
pub async fn enable_token<Token, Params>(
    ctx: MmArc,
    req: EnableTokenRequest<Params>,
) -> Result<Token::ActivationResult, MmError<EnableTokenError>>
where
    Token: TokenActivationOps + Clone,
    Params: ActivationParamsTrait,
    EnableTokenError: From<Token::ActivationError>,
    (Token::ActivationError, EnableTokenError): NotEqual,
{
```
and create 2 wrappers as below
```rust
pub async fn enable_erc20_token(
    ctx: MmArc,
    req: EnableTokenRequest<Erc20TokenActivationRequest>,
) -> Result<EthTokenInitResult, MmError<EnableTokenError>> {
    enable_token::<EthCoin>(ctx, req).await
}

pub async fn enable_nft(
    ctx: MmArc,
    req: EnableTokenRequest<NftActivationRequest>,
) -> Result<EthTokenInitResult, MmError<EnableTokenError>> {
    enable_token::<EthCoin>(ctx, req).await
}
```
Where `Erc20TokenActivationRequest` and `NftActivationRequest` implement `ActivationParamsTrait`. The above code is just a demonstration and not the real code, this can also solve the backward compatibility issues. If this requires too much changes, you can do it in the next PR.

Reply:
Since the above to work will require a big refactor, I think a better way is to rename the old `enable_token` method to `enable_fungible_token` and create another `enable_nft` method with it's own `NFTActivationOps`. We can decide how to best tackle this later since this whole refactor is too much to handle when we are trying to finish reviews for this PR. We might even find a better refactor later, but I am leaning to this last suggestion since it will decouple NFT enabling logic from Fungible tokens one.

Reply:
>  if we passed erc20 activation params to enable_nft it will work. 

Sorry, I didnt get you why it should work :( 
Lets use `#[serde(untagged)]` in enum `EthTokenActivationParams`  and change the enumirations order, so serde can parse types correctly:
PS: wthout order changing, as `required_confirmations` is Optional, serder always will treat activation params as `Erc20(Erc20TokenActivationRequest)` varian.

```rust
#[serde(untagged)]
pub enum EthTokenActivationParams {
    Nft(NftActivationRequest),
    Erc20(Erc20TokenActivationRequest),
}
```

**Request - enable nft with ERC-20 activation param**
```json
{
  "userpass": "'$USERPASS'",
  "method": "enable_nft",
  "mmrpc": "2.0",
  "params": {
    "ticker": "NFT_MATIC",
    "activation_params": {
      "required_confirmations": 3
    }
  }
}
```
**Response**
```json
{
  "mmrpc": "2.0",
  "error": "Mismatched protocol info for ERC-20",
  "error_path": "token.eth_tokens_activation",
  "error_trace": "token:137] eth_tokens_activation:152]",
  "error_type": "Internal",
  "error_data": "Mismatched protocol info for ERC-20",
  "id": null
}
```

**Request - enable nft with correct parameter**
```json
{
  "userpass": "'$USERPASS'",
  "method": "enable_nft",
  "mmrpc": "2.0",
  "params": {
    "ticker": "NFT_MATIC",
    "activation_params": {
      "url": "https://moralis-proxy.komodo.earth"
    }
  }
}
```
**Response**
```json
{"mmrpc":"2.0","result":{"nfts":{"0x48c75fbf0452fa8ff2928ddf46b0fe7629cca2ff,5":{"token_address":"0x48c75fbf0452fa8ff2928ddf46b0fe7629cca2ff","token_id":"5","chain":"POLYGON","contract_type":"ERC1155","amount":"1"},"0xc06a6ab4c2b0ed14ed5692adf324a9a463259c43,0":{"token_address":"0xc06a6ab4c2b0ed14ed5692adf324a9a463259c43","token_id":"0","chain":"POLYGON","contract_type":"ERC1155","amount":"1"},"0x22d5f9b75c524fec1d6619787e582644cd4d7422,201":{"token_address":"0x22d5f9b75c524fec1d6619787e582644cd4d7422","token_id":"201","chain":"POLYGON","contract_type":"ERC1155","amount":"3000000000000000000"}},"platform_coin":"MATIC"},"id":null}
```

Reply:
@shamardy 
in this commit https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049/commits/1fee1cfe3b652785424fd7b28602b54a2d80e15d I added eth token activation to this
```rust
/// Represents the parameters required for activating either an ERC-20 token or an NFT on the Ethereum platform.
#[derive(Clone, Deserialize)]
#[serde(untagged)]
pub enum EthTokenActivationParams {
    Nft(NftActivationRequest),
    Erc20(Erc20TokenActivationRequest),
}
```

I suppose backward compatibility was fixed.

Please, could you clarify this
>I think a better way is to rename the old enable_token method to enable_fungible_token and create another enable_nft method with it's own NFTActivationOps.

If you suggest to rename `enable_token` function in `TokenActivationOps` trait and move nft activation logic to `NFTActivationOps` trait with `async fn enable_nft` function, then I dont think its a good idea, as we are trying to make architecture more generic.

As for this note: [Aren't those basically the same method but with 2 different aliases..](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1501171050) - I feel this is over complicated. 
I believe its enough to refactor `EthTokenActivationParams` type to save backward compatibility, plus add default & optional `nft Url` support if we want it.

I feel like there's been some desynchronization in our conversation because we seem to have a different idea about the default NFT URL implementation.

I suggest to leave non-optional `url` param for nft activation. 
We can discuss do we actually need to have default url to enable nft and fetch nft info + add providers next sprint.
If we really want to have it, then we should refactor all nft requests where we require `url` parameter. Also in this comment ["If we want to use #[serde(untagged)] and avoid UB, then right now I see two ways.."](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1501834040) in the second solution I provided example how we can handle `url: Option<Url>` and default url (I treated it as default provider).

Reply:
> If you suggest to rename enable_token function in TokenActivationOps trait and move nft activation logic to NFTActivationOps trait with async fn enable_nft function, then I dont think its a good idea, as we are trying to make architecture more generic.

Well, we can leave it as it is for now and it can be refactored later if needed when adding new NFT protocols. As discussed in DM, it might be needed to decouple this in the future for different protocols (e.g. UTXOs) since `enable_token` and `enable_platform_coin_with_tokens` is used for any coin that supports tokens (Tendermint, EVM, BCH).